gh-132983: Introduce `compression` package and move `_compression` module

emmatyping opened the pull request:
This is part 1 of the [Implementation Plan](https://github.com/python/cpython/issues/132983#issue-3021300364) for PEP 784.

This change:
- introduces a `compression` package and re-exports existing modules under that package.
- Moves `Lib/_compression.py` to `Lib/compression/_common/streams.py` and updates imports to match.

I added `skip news` as I'd like to write a holistic NEWS/What's New entry once the entire implementation has landed. If people think each PR should have NEWS I can write something up.

<!-- gh-issue-number: gh-132983 -->
* Issue: gh-132983
<!-- /gh-issue-number -->

emmatyping committed:
Introduce `compression` package

This commit introduces the `compression` package, specified in PEP 784
to re-export the `lzma`, `bz2`, `gzip`, and `zlib` modules. Introduction
of `compression.zstd` will be completed in a future commit once the
`_zstd` module is merged.

This commit also moves the `_compression` private module to
`compression._common.streams`.

emmatyping committed:
Merge branch 'python:main' into compression_package

StanFromIreland commented on Lib/gzip.py:
```diff
@@ -14,6 +13,7 @@
 import time
 import weakref
 import zlib
+from compression._common import streams
```
This list was alphabetically sorted.

picnixz replied:
In general, we sort by `import` then by `from X import Y`.

emmatyping replied:
Going to mark this as resolved then since it is sorted per Bénédikt's comment.

StanFromIreland commented on Lib/lzma.py:
```diff
@@ -26,7 +26,7 @@
 from _lzma import *
 from _lzma import _encode_filter_properties, _decode_filter_properties  # noqa: F401
-import _compression
+from compression._common import streams
```
ditto

emmatyping replied:
I fixed this one, I think the other one is fine?

emmatyping committed:
Update stdlib modules list

Also update `generate_stdlib_module_names.py` to collect `compression`,
a namespace package.

emmatyping committed:
Sort lzma import in alphabetical order

picnixz commented on Lib/bz2.py:
```diff
@@ -12,7 +12,7 @@
 import io
 import os
-import _compression
+from compression._common import streams
```
You can put this one after `from builtins` now

emmatyping committed:
More import sorting

picnixz reviewed (approved):
Some imports are not sorted but they are part of the test and tests imports don't follow a specific convention (unless we manage to). In general, we try to keep the convention within the same file.

gpshead commented on Lib/bz2.py:
```diff
@@ -10,9 +10,9 @@
 __author__ = "Nadeem Vawda <nadeem.vawda@gmail.com>"
 
 from builtins import open as _builtin_open
+from compression._common import streams
```
I suggest adding an `as _streams` to this so that "streams" won't be a public name in the modules we import it within.  You could also just rename streams.py to _streams.py to avoid requiring the `as ` idioms.

emmatyping replied:
Good idea!

emmatyping replied:
Since I call this out in the PEP I will also update https://github.com/python/peps/pull/4387

gpshead commented on Lib/compression/bz2/__init__.py:
could we adopt this pattern for the docstring to avoid exposing our implementation detail to people who just want the module docstring from help(bz2)?

```python
import bz2
__doc__ = bz2.__doc__
del bz2
```

emmatyping replied:
Also a good idea :)

JelleZijlstra commented on Tools/build/generate_stdlib_module_names.py:
```diff
@@ -70,8 +70,16 @@ def list_packages(names):
             continue
-        if any(package_file.endswith(".py")
-               for package_file in os.listdir(package_path)):
+        # Walk the package directory to check if the package is a namespace
```
I'd avoid using the term "namespace" here, it already led to a lot of confusion in the PEP discussion
```suggestion
        # Walk the package directory to check if there are nested packages
```

emmatyping replied:
I think in this case "namespace" is required as the reason `compression` isn't caught by default is it is *currently* a namespace package. `compression` is currently a namespace package, but we wanted to be clear it didn't need to be that way in the future in the PEP (hence the confusion).

Nested packages doesn't clarify that this is solving the case where the package is a PEP 420 namespace package in terms of layout. I guess we could say "nested packages with no modules under the top-level package"?

gpshead replied:
is there any reason **not** to add the empty `Lib/compression/__init__.py` file so it gets treated as a normal package instead of a PEP 420 style namespaces package?  We expect no code to be in the top level package and for users to always be using `from compression import thing`.

I don't _believe_ we have any namespaces packages in the stdlib today (in my quick check).

emmatyping replied:
There aren't existing namespace packages, and there's no reason we can't add `Lib/compression/__init__.py`. My only concern is if someone does `import compression` and doesn't get anything they'll be confused. I think it is fine to to add it however.

emmatyping replied:
I've added the `__init__.py` file and reverted the changes against `generate_stdlib_module_names.py`.

emmatyping committed:
Rename compression._common.streams to compression._common._streams

emmatyping committed:
Re-export existing module docstrings

emmatyping committed:
Add compression/__init__.py and revert tooling change

merwok commented on Lib/compression/gzip/__init__.py:
Is there a need for these to be mostly-empty packages rather than modules? (Lib/compression/gzip.py)

emmatyping replied:
Mostly just to keep `compression`'s directory clean, but that's not needed per se.

merwok replied:
Would you be open to simplifying this?
